Add LightOnOCR model implementation

baptiste-aubertin opened the pull request:
# What does this PR do?

Implementation of the LightonOCR model following Modular Transformers architecture.

Our model is a 1B parameter OCR model using Pixtral as the vision encoder and Qwen3 as the LLM decoder.

I still have an issue with auto configuration, which I'm not familiar with:

`🚨 Config not found for lightonocr. You can manually add it to HARDCODED_CONFIG_FOR_MODELS in utils/auto_docstring.py`


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker @Cyrilvallez
- vision models: @yonigozlan @molbap
- audio models: @eustlb @ebezzam @vasqu
- multimodal models: @zucchini-nlp
- graph models: @clefourrier

Library:

- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- continuous batching: @remi-or @ArthurZucker @McPatate
- pipelines: @Rocketknight1
- tokenizers: @ArthurZucker and @itazap
- trainer: @SunMarc
- attention: @vasqu @ArthurZucker @CyrilVallez
- model loading (from pretrained, etc): @CyrilVallez
- distributed: @3outeille @ArthurZucker
- CIs: @ydshieh

Integrations:

- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization: @SunMarc @MekkCyber
- kernels: @MekkCyber @drbh
- peft: @BenjaminBossan @githubnemo

Devices/Backends:

- AMD ROCm: @ivarflakstad
- Intel XPU: @IlyasMoutawwakil
- Ascend NPU: @ivarflakstad 

Documentation: @stevhliu

Research projects are not maintained and should be taken as is.

 -->

baptiste-aubertin committed:
Add LightOnOCR model implementation

baptiste-aubertin committed:
fix modular docstring error

baptiste-aubertin committed:
Improve LightOnOCR documentation and exports

baptiste-aubertin committed:
Rename LightOnOCR multi-modal projector to vision projection and add tests

baptiste-aubertin committed:
fix load without lmhead in safetensor

baptiste-aubertin committed:
temp

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+            "rope_theta": 10000,
+        },
+        text_config: Optional[dict[str, Any]] = {
+            "architectures": ["Qwen3ForCausalLM"],
```
we should not use mutable defaults, it can have weird side effects if the values are changed in place somewhere down the lines. Instead let's move this to the `__init__` as `if text_config is None: text_config = {default_config_here}`

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+        },
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
```
let's move to the end, after subconfigs are initialized. That way we will dispatch correct attention implementation for subconfigs in the super call

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+    def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
+        """Load config from pretrained model."""
+        config_dict, kwargs = cls.get_config_dict(pretrained_model_name_or_path, **kwargs)
+        return cls.from_dict(config_dict, **kwargs)
```
base class methods can handle nested configs, so should not be needed. Prob it's failing now because the config class needs `cls.sub_config: dict[str, Any]`

https://github.com/huggingface/transformers/blob/354567d955fbc5fbd70fc841b7a7bcc654bea3f1/src/transformers/models/clip/configuration_clip.py#L273

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+        return cls.from_dict(config_dict, **kwargs)
+
+
+class PixtralProcessorKwargs(ProcessingKwargs, total=False):
```
ig a typo, `LightOnOCRProcessorKwargs` :)

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+        chat_template=None,
+        image_token="<|image_pad|>",
+        image_break_token="<|vision_pad|>",
+        image_end_token="<|vision_end|>",
```
these are better obtained directly from the tokenizer's special attributes, as `tokenizer.image_token_id`. Do we use the same weights as official one here or do we have to convert?

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+
+        patch_size = self.patch_size * self.spatial_merge_size
+        if images is not None:
+            output_kwargs["images_kwargs"]["patch_size"] = patch_size
```
not very fan of it. Ideally the saved config has the correct `image_processor.patch_size` and we can also set defaults for images kwargs in the `TypedDict` above. Force-setting image kwargs from within processor doesn't look good

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+    """
+)
+class LightOnOCRText(Qwen3Model):
+    pass
```
having only this layer is enough, modular will copy its dependencies automatically

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+    def __init__(self, config: LightOnOCRConfig):
+        super().__init__(config)
+
+        self.vision_encoder = LightOnOCRVision(config.vision_config)
```
`Model._from_config` so that it can initialize sub-models with their own `dtypes` if needed. Same for text model

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+        pixel_values: Optional[torch.Tensor] = None,
+        image_sizes: Optional[list[tuple[int, int]]] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+        **kwargs,
```
some args missing

```
        position_ids: Optional[torch.LongTensor] = None,
        past_key_values: Optional[Cache] = None,
        cache_position: Optional[torch.LongTensor] = None,
        use_cache: Optional[bool] = None,
```

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+                image_mask = (input_ids == self.config.image_token_id).unsqueeze(-1).expand_as(inputs_embeds)
+
+                # Replace image tokens with visual embeddings using masked_scatter
+                inputs_embeds = inputs_embeds.masked_scatter(image_mask, projected_visual)
```
let;s move it one indent back. Pixel value preparation does not depend on whether we have input ids or embeddings

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+                projected_visual = projected_visual.to(inputs_embeds.dtype)
+
+                # Create mask for image tokens
+                image_mask = (input_ids == self.config.image_token_id).unsqueeze(-1).expand_as(inputs_embeds)
```
we can use `self.get_placeholder_mask` as in llava

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+                projected_visual = self.vision_projection(visual_features.squeeze(0), image_sizes)
+
+                # Convert to same dtype
+                projected_visual = projected_visual.to(inputs_embeds.dtype)
```
and device, sometimes they end up in different devices in multiGPU setting

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+
+        # First generation step: process vision encoder if pixel_values provided
+        if inputs_embeds is None and pixel_values is not None and is_first_step:
+            pixel_values = pixel_values.to(dtype=self.dtype)
```
let's copy this method from existing VLMs, e.g. LLaVA

zucchini-nlp commented on tests/models/lightonocr/test_modeling_lightonocr.py:
```diff
@@ -0,0 +1,204 @@
+
+
+@require_torch
+class LightOnOCRModelTest(unittest.TestCase):
```
we need `ModelTestMixin` and `GeneraionTestMixin` to run basic tests on the model. You can take a look at tests in other VLMs

zucchini-nlp commented on tests/models/lightonocr/test_modeling_lightonocr.py:
```diff
@@ -0,0 +1,204 @@
+
+
+if __name__ == "__main__":
+    unittest.main()
```
same

zucchini-nlp reviewed (commented):
Hey @baptiste-aubertin , thanks a lot for working on it! Nice PR, and I left comments mostly to push the model for better standardization. Lmk when you need a second review

baptiste-aubertin committed:
Refactor LightOnOCR config to use sub_configs pattern

baptiste-aubertin committed:
rename processor kwargs

baptiste-aubertin committed:
Refactor LightOnOCR processor to use effective patch size

Calculate effective_patch_size during initialization and use it throughout
the processor. Update ProcessorKwargs defaults to include patch_size in
images_kwargs. Remove redundant model_input_names property.

baptiste-aubertin committed:
Improve LightOnOCR generation support with proper KV cache handling

baptiste-aubertin committed:
add modeling tests and compile modular

baptiste-aubertin committed:
Clean up LightOnOCR code and remove unused variables

Remove unused image_features variable and model_input_names property

baptiste-aubertin committed:
Add LightOnOCR documentation and test improvements

Add model documentation page with config and class references. Update toctree to include LightOnOCR entry. Clean up test formatting and add vision/text models to private model exceptions.

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,792 @@
+    @num_hidden_layers.setter
+    def num_hidden_layers(self, value):
+        """Set num hidden layers in text config."""
+        self.text_config.num_hidden_layers = value
```
Shouldn't need to set all these properties like this. Instead, you should access the attributes through `config.text_config.attribute_name` in the modular code, that will make the config cleaner too

```suggestion
```

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,792 @@
+    The language model of LightOnOCR, based on Qwen3 architecture.
+    """
+)
+class LightOnOCRText(Qwen3Model):
```
```suggestion
class LightOnOCRTextModel(Qwen3Model):
```

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,792 @@
+
+__all__ = [
+    "LightOnOCRPreTrainedModel",
+    "LightOnOCRText",
```
```suggestion
    "LightOnOCRTextModel",
```

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,792 @@
+        cache_position: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
```
These are not used anymore, we use the decorator     `@check_model_inputs()` to possibly intercept states. Overall four decorators are widely used in the library and simplify modeling a lot, `auto_docstring`, `can_return_tuple`, `check_model_inputs` and `_can_record_outputs`, make sure you're making use of them!

```suggestion
```

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,792 @@
+        self.model.set_input_embeddings(value)
+
+    def get_output_embeddings(self):
+        return self.lm_head
```
This is the default in `PreTrainedModel` so should be safe to remove as well (not 100% sure but 85% sure?)
```suggestion
```

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,724 @@
+            image_grid = image_tokens.view(h, w, d).permute(2, 0, 1).unsqueeze(0)
+            # shape [1, d, h, w] -> [h // sms * w // sms, d * sms**2]
+            # sms = spatial_merge_size
+            # note(staghado): when h or w is not divisible by sms, the last row/column will be ignored??
```
+1

molbap commented on tests/models/lightonocr/test_modeling_lightonocr.py:
```diff
@@ -0,0 +1,524 @@
+        """
+        # This test would be similar to LLaVA's integration tests
+        # but would require an actual pretrained LightOnOCR checkpoint
+        pass
```
Let's remove this for now

molbap reviewed (commented):
Hey @baptiste-aubertin, continued the review! Also for naming, the model should be better named `LightOnOcr` rather than `..OCR` but that's a casing nit. For the tests not passing, adding the Text and Vision models to `utils/check_repo` list `IGNORE_NON_TESTED` should be enough.
I recommend rebasing and running the linter again, from your branch:
```bash
git checkout main
git pull
git checkout -
git merge main
pip install -e .[quality]
make fixup
```
and that should get rid of the code quality issues.

baptiste-aubertin committed:
Refactor LightOnOCR to use standardized RopeParameters and consolidate shared components

baptiste-aubertin committed:
Rename LightOnOCR model classes and fix config parameter naming

- Rename LightOnOCRText -> LightOnOCRTextModel and LightOnOCRVision -> LightOnOCRVisionModel
- Fix parameter naming: image_token_index -> image_token_id
- Set tie_word_embeddings default to False
- Add special case for inherited Qwen3Config attributes in LightOnOCRTextConfig

baptiste-aubertin committed:
Add missing parameter documentation for LightOnOCR config

baptiste-aubertin committed:
Simplify LightOnOCR forward methods with decorators and fix loss function call

baptiste-aubertin committed:
Reorganize LightOnOCR components to place vision before text and remove debug print

baptiste-aubertin committed:
fixup

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,691 @@
+    """
+)
+class LightOnOCRVisionModel(PixtralVisionModel):
+    config_class = LightOnOCRVisionConfig
```
I see, indeed this can break due to naming conventions on used models. However I'm not sure it's breaking right now, can you elaborate a bit? If it does break you can always go back to redefining pass classes like this (
```python
class LightOnOCRVisionRMSNorm(PixtralRMSNorm):
    pass


class LightOnOCRVisionMLP(PixtralMLP):
    pass


class LightOnOCRVisionRotaryEmbedding(PixtralRotaryEmbedding):
    pass


class LightOnOCRVisionAttention(PixtralAttention):
    def __init__(self, config):
        super().__init__(config)
        self.num_key_value_groups = config.num_key_value_groups


class LightOnOCRVisionAttentionLayer(PixtralAttentionLayer):
    pass


class LightOnOCRVisionTransformer(PixtralTransformer):
    pass
```

BTW I added the num_key_value_groups attribute because I think it might be causing you trouble in current tests. It's expecting to find this attribute in generic tests and since this is just MHA it's absent

staghado replied:
Sadly this does not work bcz both vision encoder and LM decoder are calling the same `eager_attention_forward` and modular only writes one to the modeling file.

baptiste-aubertin replied:
Hi, thanks for the answer.
I tried to reimport all the Pixtral modules like that, but both Pixtral and Qwen3 attentions are trying to use the same `eager_attention_forward` function. Do I need to copy all the Qwen and Pixtral implementations into my modular file and rename the eager attention functions?

molbap reviewed (commented):
I think you're right, some of the current tests failing are indeed because the text decoder end up using `eager_attention_forward` from Pixtral module which uses MHA, but the Qwen3 module needs GQA/repeating key values. Here the repeat should be 2, so you get a factor 2 issue in your matmuls 😬 To circumvent it you can rewrite the needed `eager_attention_forward` with `repeat_kv` so you'll have GQA in the decoder as expected. I outlined changes in the vision modules to silo the inheritance a bit, and similar ones might be needed in the text modules. Let me know and ping me when you've done a new pass!

staghado committed:
Fix image token expansion logic in Processor

baptiste-aubertin committed:
Copy pixtral attention to have both pixtral and qwen eager attention forward

baptiste-aubertin committed:
remove LightOnOCRTextPreTrainedModel from modular to be able to return attention

baptiste-aubertin committed:
Support both tensor and list formats for image_sizes parameter

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,847 @@
+        ]
+
+        tokens_per_image = [h * w for h, w in image_sizes_in_patches]
+        d = image_features.shape[-1]
```
nit, no single-letter variables

molbap reviewed (commented):
Looking much better! Left what I think are my last comments, once addressed we'll ping core maintainer for final review & merge :partying_face:

baptiste-aubertin committed:
Update tests/models/lightonocr/test_processor_lightonocr.py

Co-authored-by: Pablo Montalvo <39954772+molbap@users.noreply.github.com>

baptiste-aubertin committed:
Update docs/source/en/model_doc/lightonocr.md

Co-authored-by: Pablo Montalvo <39954772+molbap@users.noreply.github.com>

baptiste-aubertin committed:
Move image_sizes tensor conversion from model to processor

baptiste-aubertin committed:
Simplify weight initialization to use uniform text_config initializer_range

baptiste-aubertin committed:
rename 1 letter vars

baptiste-aubertin committed:
Get image special tokens from tokenizer attributes in processor

baptiste-aubertin committed:
Return BaseModelOutputWithPast from LightOnOCRModel forward

baptiste-aubertin committed:
Add chat template to LightOnOCR processor test setup

baptiste-aubertin committed:
rm get_output_embeddings from LightOnOCRForConditionalGeneration (not needed)

baptiste-aubertin committed:
Add OCR integration test for LightOnOCR model

Tests model can perform OCR on real receipt image and extract expected text

baptiste-aubertin committed:
Fix device/dtype handling in LightOnOCR vision processing

baptiste-aubertin committed:
Add TransformersKwargs type hints to LightOnOCR forward methods

baptiste-aubertin committed:
Make torch imports conditional and use _from_config for LightOnOCR sub-models

baptiste-aubertin committed:
Set patch_size at runtime instead of modifying class defaults in LightOnOCR processor

molbap commented:
run-slow: lightonocr

molbap commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,848 @@
+        self,
+        images: Optional[ImageInput] = None,
+        text: Union[TextInput, PreTokenizedInput, list[TextInput], list[PreTokenizedInput]] = None,
+        **kwargs,
```
let's type the kwargs with `Unpack[LightOnOCRProcessorKwargs]`

molbap commented on tests/models/lightonocr/test_modeling_lightonocr.py:
```diff
@@ -0,0 +1,678 @@
+        # Check that the model correctly extracted the date from the receipt
+        self.assertIn(
+            "25/12/2018", decoded_output, "Model should extract the date '25/12/2018' from the receipt image"
+        )
```
For that model, don't we have a deterministic idea of what its output is? To be more robust to regressions, just compare to the full expected output, no?

molbap reviewed (approved):
Left last nits comments + a non-nit comment on testing as we only test on part of the output for now in the IntegrationTest, the rest seems copacetic :) well done, ccing to core reviewer

molbap commented:
There's a couple errors on the `slow tests` I just ran: with 

```python
FAILED tests/models/lightonocr/test_modeling_lightonocr.py::LightOnOCRForConditionalGenerationIntegrationTest::test_model_can_generate_without_images - AttributeError: 'LightOnOCRConfig' object has no attribute 'vocab_size'
FAILED tests/models/lightonocr/test_modeling_lightonocr.py::LightOnOCRForConditionalGenerationIntegrationTest::test_model_forward_with_images - AttributeError: 'LightOnOCRConfig' object has no attribute 'vocab_size'
```

so just a config key missing/a wrong access IMO. The other one which is suspicious is

```python
    def vision_apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
        """Applies Rotary Position Embedding to the query and key tensors.
        .......
        """
        cos = cos.unsqueeze(unsqueeze_dim)
        sin = sin.unsqueeze(unsqueeze_dim)
>       q_embed = (q * cos) + (vision_rotate_half(q) * sin)
E       RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!
```

which is on a multi-GPU setting only but means that a q tensor should be created on a safer device, or at least within this util function to be safe.

baptiste-aubertin committed:
type kwargs

baptiste-aubertin commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,848 @@
+
+        # Since we use packing, if flash_attention_2 is selected we rely on position_ids
+        if self.config._attn_implementation == "flash_attention_2":
+            kwargs["position_ids"] = kwargs["position_ids"].to(hidden_states.device, non_blocking=True)
```
https://github.com/huggingface/transformers/blob/6fb6d3c0fbce052bffa061ad41e1a20e7558e73a/src/transformers/models/pixtral/modeling_pixtral.py#L245

baptiste-aubertin committed:
Remove loocr forward comments

baptiste-aubertin committed:
Add vocab_size property and fix image_token_id in LightOnOCR

- Add vocab_size property to LightOnOCRConfig that delegates to text_config
- Fix test parameter name from image_token_index to image_token_id
- Add Unpack type hint to processor __call__ kwargs
- Remove unnecessary comments from modeling forward method

baptiste-aubertin committed:
Add vocab_size setter to LightOnOCR configuration

baptiste-aubertin committed:
Fix device mismatch in vision rotary embeddings and optimize test image sizes

baptiste-aubertin committed:
Improve LightOnOCR integration test with similarity-based output validation

baptiste-aubertin committed:
Enable flex attention

baptiste-aubertin committed:
Enable flex attention

zucchini-nlp commented:
run-slow: lightonocr

zucchini-nlp commented on docs/source/en/model_doc/lightonocr.md:
```diff
@@ -0,0 +1,62 @@
+# LightOnOCR
+
+LightOnOCR is a multimodal model designed for optical character recognition (OCR) tasks. It combines a vision encoder for processing document images with a text decoder for generating text sequences.
+
```
maybe we can also add a link to blogpost https://huggingface.co/blog/lightonai/lightonocr since there is no paper

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,844 @@
+
+    @vocab_size.setter
+    def vocab_size(self, value):
+        self.text_config.vocab_size = value
```
same, ideally it should be obtained with `config.get_text_config().vocab_size`. We've been trying to delete high level getters

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,844 @@
+            if not isinstance(image_inputs["image_sizes"], torch.Tensor):
+                image_inputs["image_sizes"] = torch.tensor(image_inputs["image_sizes"])
+            else:
+                image_inputs["image_sizes"] = image_inputs["image_sizes"].clone()
```
didn't really get this part. I'd assume that image processor returns in tensors already if requested, so maybe there's smth to fix in the image processor

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,844 @@
+            List of image feature tensors, one per image
+        """
+        # Convert image_sizes tensor to list of tuples for compatibility with vision encoder
+        image_sizes_list = [(int(h), int(w)) for h, w in image_sizes]
```
i am surprised to see this. The pixtral encoder handles tensors as inputs iirc because it is used in other models  in similar way

zucchini-nlp reviewed (commented):
Hye @baptiste-aubertin , While we're waiting on core maintainer's review, I left a few comments and questions below, mostly about things that looks redundant to me. LMK what you think and if we can delete those?

baptiste-aubertin committed:
Loocr description with blogpost

baptiste-aubertin committed:
redundant tie_word_embeddings

baptiste-aubertin committed:
remove architecture from default config

baptiste-aubertin committed:
vocab_size accessors

baptiste-aubertin committed:
remove useless tensor conversion

baptiste-aubertin committed:
remove useless conversion

baptiste-aubertin committed:
move dtype conversion to after image feature extraction

baptiste-aubertin commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,844 @@
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
```
yes ! i works if i return the output of the language model but it doesn't if i just remove them from BaseModelOutputWithPast

baptiste-aubertin committed:
remove useless stuff

baptiste-aubertin commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,844 @@
+        # These should be set on the tokenizer before creating the processor
+        self.image_token = getattr(tokenizer, "image_token", "<|image_pad|>")
+        self.image_break_token = getattr(tokenizer, "image_break_token", "<|vision_pad|>")
+        self.image_end_token = getattr(tokenizer, "image_end_token", "<|vision_end|>")
```
If it's okay for you we will do that with the vllm harmonisation after ?

molbap commented:
run-slow: lightonocr

baptiste-aubertin committed:
fixup

baptiste-aubertin committed:
export text and vision config classes

baptiste-aubertin committed:
refactor(lightonocr): remove unused weight initialization and fix tied weights mapping #0

- Remove custom _init_weights methods (handled by base class)
- Update _tied_weights_keys to dict format with explicit mapping
- Update documentation date

baptiste-aubertin commented:
A test is not passing but it not seems to be related with our model. Could you please run the slow tests @molbap ?

molbap commented:
There's just these:

```bash
 FAILED tests/models/lightonocr/test_modeling_lightonocr.py::LightOnOCRForConditionalGenerationModelTest::test_model_parallelism - RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!
FAILED tests/models/lightonocr/test_modeling_lightonocr.py::LightOnOCRForConditionalGenerationIntegrationTest::test_model_can_generate_without_images - AttributeError: 'LightOnOCRConfig' object has no attribute 'vocab_size'
FAILED tests/models/lightonocr/test_modeling_lightonocr.py::LightOnOCRForConditionalGenerationIntegrationTest::test_model_forward_with_images - AttributeError: 'LightOnOCRConfig' object has no attribute 'vocab_size'
```

For the last two - thought it was solved? but for the first one it's an "actual" issue, although you should be able to fix it by either creating the tensor in the same device as the first one, or just make sure you don't have any lingering .to(device) as device_map="auto" should be able to handle the device placement!

baptiste-aubertin committed:
fix(lightonocr): fix test failures for vocab_size access and device placement #0

- Use config.text_config.vocab_size instead of config.vocab_size for composite config
- Remove explicit device placement from attention_mask and image_sizes tensors
- Allow device_map='auto' to handle device placement in model parallelism tests

baptiste-aubertin committed:
ruff

baptiste-aubertin committed:
rebase 8/12/2025

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,793 @@
+class LightOnOCRProcessor(ProcessorMixin):
+    attributes = ["image_processor", "tokenizer"]
+    image_processor_class = "AutoImageProcessor"
+    tokenizer_class = "AutoTokenizer"
```
nit: these can be removed after rebasing main. There was some refactoring recently

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,793 @@
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["LightOnOCRVisionProjector", "LightOnOCRPatchMerger"]
+    _skip_keys_device_placement = "past_key_values"
+    _supports_flash_attn_2 = True
```
the attribute is without the `_2` suffix in main :)

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,793 @@
+
+class LightOnOCRModel(LightOnOCRPreTrainedModel):
+    base_model_prefix = "model"
+    _checkpoint_conversion_mapping = {}
```
no need to set explicitly, because we don't copy from llava or other VLMs as I see

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,793 @@
+    @property
+    def vision_model(self):
+        """Alias for vision_encoder to match standard composite model naming."""
+        return self.model.vision_encoder
```
also to delete

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,793 @@
+
+        if images is not None:
+            # Like pixtral
+            output_kwargs["images_kwargs"]["patch_size"] = self.effective_patch_size
```
again, this is not the best way to change image processing attributes. We should save an image processor with correct `patch_size` in config and let it be loaded/used

zucchini-nlp reviewed (approved):
The tests are all passing 🎉 

I see you rebased main branch. We've some changes recently in processing/modeling code for the v5 release, I left comment on what needs to be updated and deleted

I see comments from @molbap were resolved, so this should be ready for core maintainer's review

baptiste-aubertin committed:
rebase 09/12/2025

baptiste-aubertin committed:
review zucchini

baptiste-aubertin committed:
review zucchini

baptiste-aubertin committed:
rebase 10/12/2025

Cyrilvallez commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,752 @@
+from collections.abc import Callable
+from typing import Any, Optional, Union
```
Missing license at the top here

Cyrilvallez commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,752 @@
+        # These should be set on the tokenizer before creating the processor
+        self.image_token = getattr(tokenizer, "image_token", "<|image_pad|>")
+        self.image_break_token = getattr(tokenizer, "image_break_token", "<|vision_pad|>")
+        self.image_end_token = getattr(tokenizer, "image_end_token", "<|vision_end|>")
```
As mentionned previously, should be set directly in the tokenizer, not here! To remove

Cyrilvallez commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,752 @@
+        self.image_token_id = tokenizer.convert_tokens_to_ids(self.image_token)
+        self.image_break_token_id = tokenizer.convert_tokens_to_ids(self.image_break_token)
+        self.image_end_token_id = tokenizer.convert_tokens_to_ids(self.image_end_token)
+
```
If they are in the tokenizer, you should be able to simply do `tokenizer.image_token_id` to get it

Cyrilvallez commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,752 @@
+    _supports_flash_attn = True
+    _supports_sdpa = True
+    _supports_flex_attn = True
+    _no_split_modules = ["LightOnOCRVisionAttentionLayer"]
```
We can merge both, no need to have 2 as they are similar

Cyrilvallez commented on tests/models/lightonocr/test_modeling_lightonocr.py:
```diff
@@ -0,0 +1,676 @@
+
+    @unittest.skip("Pixtral does not support attention interfaces.")
+    def test_flex_attention_with_grads(self):
+        pass
```
All those have no reason to be skipped -> to remove

Cyrilvallez reviewed (commented):
Hey! Nice work! Are a few comments to still address before merging!
Also, let's comply to our conventions of `snake_case` and `CamelCase`, and let's rename `lightonocr` -> `lighton_ocr` and `LightOnOCR` -> `LightOnOcr` everywhere!

baptiste-aubertin committed:
refactor(lighton_ocr): fix naming conventions to use snake_case and proper CamelCase #0

- Rename model identifier from 'lightonocr' to 'lighton_ocr' (snake_case)
- Update class names from 'LightOnOCR*' to 'LightOnOcr*' (proper CamelCase)
- Update all auto mappings, tests, and documentation accordingly

baptiste-aubertin committed:
style(lighton_ocr): remove unnecessary import guards for torch and vision #0

baptiste-aubertin committed:
style(lighton_ocr): remove unnecessary pass statement from LightOnOcrVisionConfig #0

baptiste-aubertin committed:
refactor(lighton_ocr): consolidate RMSNorm classes and use PixtralRMSNorm base #0

baptiste-aubertin committed:
refactor(lighton_ocr): import rotary pos emb functions from pixtral instead of redefining #0

- Remove duplicate vision_rotate_half and vision_apply_rotary_pos_emb functions
- Import apply_rotary_pos_emb from pixtral modeling
- Consolidate rotate_half/apply_rotary_pos_emb in generated modeling file

baptiste-aubertin committed:
refactor(lighton_ocr): remove unused LightOnOcrVisionPreTrainedModel class #0

- Remove redundant VisionPreTrainedModel class that was not used
- Add LightOnOcrVisionAttentionLayer to _no_split_modules in main PreTrainedModel

baptiste-aubertin committed:
refactor(lighton_ocr): simplify LightOnOcrAttention and clarify docstring #0

- Remove redundant __init__ that only called super()
- Update docstring to explain why class exists (avoids eager_attention_forward collision with Qwen3)

baptiste-aubertin committed:
test(lighton_ocr): remove unnecessary skipped test methods #0

baptiste-aubertin committed:
refactor(lighton_ocr): remove use_sliding_window and max_window_layers from config #0

- Use del in __init__ to explicitly remove inherited attrs from Qwen3Config
- Remove LightOnOCRTextConfig from check_config_attributes.py exception list
- Fix rms_norm_eps type annotation from int to float

baptiste-aubertin committed:
fix make fixup

baptiste-aubertin committed:
docs(lighton_ocr): add docstring to LightOnOcrTextConfig and clean up check_repo #0

- Add configuration docstring with all parameters to LightOnOcrTextConfig
- Consolidate duplicate comments in PRIVATE_MODELS
- Remove redundant entries from IGNORE_NON_TESTED and IGNORE_NON_AUTO_CONFIGURED

baptiste-aubertin committed:
chore(lighton_ocr): update copyright headers to LightOn Team #0

baptiste-aubertin committed:
refactor(lighton_ocr): clean up model files and add license headers #0

- Add Apache 2.0 license headers to generated files
- Remove unused embedding getter/setter methods from ForConditionalGeneration
- Clean up LightOnOcrTextConfig docstring and remove Qwen references

baptiste-aubertin commented on src/transformers/models/lighton_ocr/modular_lighton_ocr.py:
```diff
@@ -0,0 +1,752 @@
+        return self.language_model.get_input_embeddings()
+
+    def set_input_embeddings(self, value):
+        self.language_model.set_input_embeddings(value)
```
Hi @Cyrilvallez @zucchini-nlp, I tried removing these methods but the tests fail with:                                
                                                                                                                        
  NotImplementedError: `get_input_embeddings` not auto‑handled for LightOnOcrModel; please override in the subclass.    
                                                                                                                        
  I checked the auto-handling in PreTrainedModel.get_input_embeddings() (modeling_utils.py:1066-1085) and it only looks 
  for:                                                                                                                  
  - self.embed_tokens                                                                                                   
  - self.embeddings.embed_tokens                                                                                        
  - self.model.embed_tokens                                                                                             
  - self.base_model.get_input_embeddings()                                                                              
                                                                                                                        
  It doesn't check self.language_model, which is what we use here.                                                      
                                                                                                                        
  I also checked other VLMs and they all still have these methods:                                                     
  - Qwen3VLModel (modeling_qwen3_vl.py:938)                                                                             
  - Llama4VisionModel (modeling_llama4.py:1184)                                                                         
  - VideoLlama3Model (modeling_video_llama_3.py:534)                                                                    
  - Ovis2Model (modeling_ovis2.py:517)    
  
  Maybe i misunderstood something about this ?

zucchini-nlp replied:
Ah right, the language model is not a base model here. Makes sense. Let's remove the `self.vision_model` attribute though, it was cleaned up recently for all models. The suggested way is to call `self.get_encoder(modality="image")`

baptiste-aubertin committed:
refactor(lighton_ocr): simplify processor token access and test setup #0

- Access special tokens directly from tokenizer attributes instead of getattr with defaults
- Simplify test setup to use model_id and inherited ProcessorTesterMixin methods
- Fix return types test to handle fast image processor limitations

zucchini-nlp commented on src/transformers/models/lightonocr/modular_lightonocr.py:
```diff
@@ -0,0 +1,752 @@
+        cos, sin = position_embeddings
+        query_states, key_states = vision_apply_rotary_pos_emb(query_states, key_states, cos, sin, unsqueeze_dim=0)
+
+        attention_interface: Callable = vision_eager_attention_forward
```
if it is only in `eager_attention_forward` and GQA, we can set a dummy attr for `num_kv_groups=1` in the vision attention module

It allows us to re-use the same eager fn for vision and text attentions

zucchini-nlp commented:
@baptiste-aubertin replied to the two comments and rebased main. Can you make sure that tests are passing and run `make fix-copies` to adjust modular conversion?

baptiste-aubertin committed:
refactor(lighton_ocr): unify attention functions and fix buffer registration #0

- Remove duplicate vision_eager_attention_forward, reuse eager_attention_forward from Qwen3
- Add num_key_value_groups attribute for GQA compatibility
- Register original_inv_freq as buffer instead of plain attribute

baptiste-aubertin committed:
refactor(lighton_ocr): remove vision_model property alias #0

baptiste-aubertin committed:
docs(lighton_ocr): add usage example and update release date #0

baptiste-aubertin committed:
rebase 12/01/26

Cyrilvallez committed:
Merge branch 'main' into main

Cyrilvallez commented on src/transformers/models/lighton_ocr/modular_lighton_ocr.py:
```diff
@@ -0,0 +1,779 @@
+
+
+class LightOnOcrModel(LightOnOcrPreTrainedModel):
+    base_model_prefix = "model"
```
This is a base model, so it does not have a prefix
```suggestion
    base_model_prefix = ""
```

Cyrilvallez commented on tests/models/lighton_ocr/test_modeling_lighton_ocr.py:
```diff
@@ -0,0 +1,620 @@
+    _is_composite = True
+    test_head_masking = False
+    test_pruning = False
+    test_torchscript = False
```
These 3 do not exist anymore, to remove

Cyrilvallez commented on tests/models/lighton_ocr/test_modeling_lighton_ocr.py:
```diff
@@ -0,0 +1,620 @@
+            elif model_class == LightOnOcrModel:
+                self.assertTrue(hasattr(model, "vision_encoder"))
+                self.assertTrue(hasattr(model, "language_model"))
+                self.assertTrue(hasattr(model, "vision_projection"))
```
Do not exist anymore, to remove!

Cyrilvallez reviewed (commented):
Hey! Very very nice! Only a few minor final comments, then we can merge! Great work! 🚀🚀🤗

baptiste-aubertin committed:
Update docs/source/en/model_doc/lighton_ocr.md

Co-authored-by: Cyril Vallez <cyril.vallez@gmail.com>

baptiste-aubertin committed:
review cyril

baptiste-aubertin committed:
review cyril

baptiste-aubertin committed:
review cyril

baptiste-aubertin committed:
Remove test.py from version control

Cyrilvallez committed:
Merge branch 'main' into main

baptiste-aubertin committed:
apply modular

Cyrilvallez committed:
update years everywhere it was not updated

baptiste-aubertin committed:
fix date

Cyrilvallez committed:
remove Attention forward implem

Cyrilvallez committed:
Fix all Vision prefixes instead of no prefix

Cyrilvallez committed:
move tying to main config

Cyrilvallez committed:
fix

Cyrilvallez committed:
add to all

Cyrilvallez committed:
immensely simplify

baptiste-aubertin committed:
fix test

Cyrilvallez committed:
revert check_repo

Cyrilvallez reviewed (approved):
All right, as discussed offline I simplified a looot by simply inheriting from Mistral3!
Everything is now ready, congrats! Gonna merge it! 🤗🚀
